ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
-            // Shift operand down the the LSB.
-            let op = self.0 >> self.short_operand_bit_off(index);
-            // Then mask it out.
-            Operand::Short(ShortOperand(op & SHORT_OPERAND_MASK))
-        } else {
-            todo!()
+impl CallInstruction {
+    pub(crate) fn new(m: &mut Module, target: FuncIndex, args: &[Operand]) -> CallInstruction {
+        let mut arg1 = PackedOperand::default();
+        let mut extra = ExtraArgsIndex::default();
+
+        if args.len() >= 1 {
+            arg1 = PackedOperand::new(&args[0]);
+        }
+        if args.len() >= 2 {
+            extra = m.push_extra_args(&args[1..]);
+        }
+        Self {
+            target,
+            arg1,
+            extra,
         }
     }
 
-    /// Create a new `Load` instruction.
-    ///
-    /// ## Operands
-    ///
-    /// - `<ptr>`:  The pointer to load from.
+    /// Fetch the operand at the specified index.
     ///
-    /// ## Semantics
-    ///
-    /// Return the value obtained by dereferencing the operand (which must be pointer-typed).
-    pub(crate) fn create_load(op: Operand) -> Self {
-        if op.is_short() {
-            let mut instr = Instruction::new_short(OpCode::Load);
-            instr.set_short_operand(op, 0);
-            instr
+    /// It is undefined behaviour to provide an out-of-bounds index.
+    pub(crate) fn operand(&self, m: &Module, idx: usize) -> Option<Operand> {
+        #[cfg(debug_assertions)]
+        {
+            let func = m.get_func(self.target);
+            // TRATTPLSFIX: Is there a one-liner for this?
+            if let Type::Func(ft) = m.get_type(func.tyidx) {
+                debug_assert!(ft.num_args() > idx);
+            } else {
+                panic!()
+            };
+        }
+        if idx == 0 {
+            Some(self.arg1.get())
         } else {
-            todo!();
+            Some(m.extra_args[usize::try_from(self.extra.0).unwrap() + idx - 1].clone())
         }
     }
+}
 
-    /// Create a new `LoadArg` instruction.
-    ///
-    /// ## Operands
-    ///
-    /// FIXME
-    ///
-    /// ## Semantics
-    ///
-    /// FIXME
-    pub(crate) fn create_loadarg() -> Self {
-        Instruction::new_short(OpCode::LoadArg)
-    }
-
-    /// Set the short operand at the specified index.
-    fn set_short_operand(&mut self, op: Operand, idx: u64) {
-        debug_assert!(self.is_short());
-        debug_assert!(idx < SHORT_INSTR_MAX_OPERANDS);
-        self.0 |= op.raw() << self.short_operand_bit_off(idx);
-    }
-
-    /// Returns `true` if the instruction defines a local variable.
-    pub(crate) fn is_def(&self) -> bool {
-        match self.opcode() {
-            OpCode::Load => true,
-            OpCode::LoadArg => true,
-        }
-    }
+/// Describes a function.
+#[derive(Clone, Debug)]
+pub(crate) struct Function {
+    /// The symbol name for the function.
+    name: String,
+    /// The type of the function.
+    tyidx: TypeIndex,
```

Comment:
Elsewhere `ty` has implied "index" without `idx` as a suffix. In fairness I think `idx` as a suffix makes things clearer. But whichever we choose, consistency is key.

Reply:
I think I agree, since we have different structs for types and type indices.